### PR TITLE
Refactor LSP to use vscode-jsonrpc

### PR DESCRIPTION
**Pull request by mischnic**
This makes the connection LSP server <-> Parcel reporter plugin typed by using `vscode-jsonrpc` instead of `node-ipc`. (Also also makes it individual connections, so that request-based features work as well, as opposed to just broadcasting).

With that, also add diagnostics based on symbol data (these are computed on demand, with a pull-based model when opening a file instead of being pushed on `buildSuccess` from the reporter). For that to work, the server now tries to find the matching Parcel instance via the longest common prefix between document filepath and the Parcel project root.

Bildschirm­foto 2022-12-31 um 00 24 29

Bildschirm­foto 2022-12-31 um 00 24 35

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.43s | -2.00ms |
| Cached | 313.00ms | -1.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 81.00ms | -150.00ms 🚀 |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 82.00ms | -150.00ms 🚀 |
| dist/modern/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 81.00ms | -151.00ms 🚀 |
| dist/legacy/index.html | 826.00b | +0.00b | 484.00ms | -29.00ms 🚀 |
| dist/modern/index.html | 749.00b | +0.00b | 483.00ms | -29.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 84.00ms | -145.00ms 🚀 |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 84.00ms | -146.00ms 🚀 |
| dist/modern/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 84.00ms | -145.00ms 🚀 |
| dist/legacy/index.html | 826.00b | +0.00b | 518.00ms | +32.00ms ⚠️ |
| dist/modern/index.html | 749.00b | +0.00b | 517.00ms | +31.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 9.36s | -3.00ms |
| Cached | 427.00ms | +13.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.58m | -483.00ms |
| Cached | 2.17s | +50.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.71s | -2.00ms |
| Cached | 256.00ms | -8.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/352956635585446487)**

File: packages/reporters/lsp-reporter/src/LspReporter.js [hunk 1]
```diff
@@ -1,125 +1,138 @@
...
-        pipeFilename = path.join(pathname, String(process.pid));
+        server = await createServer(SOCKET_FILE, connection => {
+          // console.log('got connection');
+          connections.push(connection);
+          connection.onClose(() => {
+            connections = connections.filter(c => c !== connection);
+          });
+
+          connection.onRequest(RequestDocumentDiagnostics, async uri => {
+            console.log('got request RequestDocumentDiagnostics', uri);
```
Comment (devongovett) [inline packages/reporters/lsp-reporter/src/LspReporter.js]: do we want to keep these logs?

File: packages/reporters/lsp-reporter/src/ipc.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+//         console.log('onRequest', f);
+
+//         setTimeout(() => {
+//           connection.sendNotification(notifyState, 'packaging');
+//         }, 200);
+//         return 123;
+//       });
+//     },
+//   );
+// })();
```
Comment (devongovett) [inline packages/reporters/lsp-reporter/src/ipc.js]: needed?

**Review by devongovett — APPROVED** [review]
lgtm, but maybe someone more familiar with the extension should review

**Review by lettertwo — APPROVED** [review]

**Pull request merged**